### PR TITLE
Add a `rustfmt.toml`

### PR DESCRIPTION
**Pull request by expenses**
See https://github.com/paritytech/substrate/issues/8981. My idea for this PR is to attempt to come up with some `rustfmt.toml` settings that fit the current style of the codebase as much as possible. I've only ran it the stuff in `utils` so far as there are less files here than elsewhere.

### CODE REVIEW

**Comment by TriplEight** [comment]
https://github.com/reviewdog/reviewdog might be of a help here

**Comment by expenses** [comment]
Needs to be run with `cargo +nightly fmt` as some settings are currently only available on nightly.

File: utils/fork-tree/src/lib.rs
Comment (expenses) [inline utils/fork-tree/src/lib.rs:275]: I don't think there's a rustfmt setting that puts the `where` on the same line as the result type unfortunately.

**Comment by expenses** [comment]
I think this is the closest style to ours that we can currently get with `cargo fmt`. I'm happy to either merge or close this PR depending on whether this is acceptable.

File: rustfmt.toml
Comment (kianenigma) [inline rustfmt.toml:6]:
```suggestion
hard_tabs = true
use_small_heuristics = "Max"
max_width = 100
comment_width = 100
chain_width = 100
merge_imports = true
reorder_imports = false
spaces_around_ranges = true
binop_separator = "Back"
```

This is what I recommend, some of them are unstable though.

Comment (shawntabrizi) [reply rustfmt.toml:6]: > reorder_imports = false

Why not true?

Comment (expenses) [reply rustfmt.toml]: Merging imports is a good idea. I think we should reorder them though, as they're unordered at the moment just because nobody (rightfully) can be bothered to do that manually. I think `chain_width` should be less than `max_width` as long chains are quite hard to read.

File: utils/frame/rpc/support/src/lib.rs
Comment (shawntabrizi) [inline utils/frame/rpc/support/src/lib.rs:55]:
```suggestion
/// #     pub struct Module<T: Config> for enum Call where origin: T::Origin {}
```

something strange with the spacing here before

Comment (expenses) [reply utils/frame/rpc/support/src/lib.rs]: Yeah, but that was weird before the formatting. I've fixed it now I think.

File: utils/browser/src/lib.rs
Comment (burdges) [inline utils/browser/src/lib.rs:156]: Just curious, can this newline be disabled?  I donno substrate's exiting style though.

File: utils/browser/src/lib.rs
Comment (burdges) [inline utils/browser/src/lib.rs:176]: I'd add some space _somewhere_ when doing this one liner myself, like `.map( |s| JsValue::from_str(&s) )..`, but that's probably not substrate's style.

File: rustfmt.toml
Comment (burdges) [reply rustfmt.toml]: I wish there was an option to order in groups, so std/core, crates.io, crates same repo, same crate.

File: utils/browser/src/lib.rs
Comment (expenses) [reply utils/browser/src/lib.rs:176]: Yeah, it's not IMO and this isn't a setting we can set afaik

File: rustfmt.toml
Comment (expenses) [reply rustfmt.toml]: > I wish there was an option to order in groups, so std/core, crates.io, crates same repo, same crate.

Yeah, it's a shame all the `sp_` imports get mixed in with the rest.

**Comment by burdges** [comment]
We should create rustfmt issues for outstanding feature requests here, like grouping imports, etc.

File: utils/browser/src/lib.rs
Comment (burdges) [reply utils/browser/src/lib.rs:176]: It looks hard to do well, but i created an issue anyways.. https://github.com/rust-lang/rustfmt/issues/4859

File: rustfmt.toml
Comment (burdges) [reply rustfmt.toml]: Added https://github.com/rust-lang/rustfmt/issues/4860

**Comment by burdges** [comment]
It maybe does not come up in substrate much, but how do you preserve rustfmt mangling a significant mixture of new lines and one liners when running rust fmt?  

It helps distinguish mathematical stuff going on in long iterator chains ala https://github.com/w3f/schnorrkel/blob/master/src/vrf.rs#L967-L969  It'd be doable by using named temporaries, except then closures break NLL.

**Comment by expenses** [comment]
> It maybe does not come up in substrate much, but how do you preserve rustfmt mangling a significant mixture of new lines and one liners when running rust fmt?

I don't think it's a problem to use `#[rustfmt::skip]` on things like that, provided it's not used _so much_ that it makes everything ugly.

File: utils/build-script-utils/src/lib.rs
Comment (drahnr) [inline utils/build-script-utils/src/lib.rs:21]: It might be preferable to avoid the re-ordering when introducing a rustfmt toml to minimize the changeset.

File: utils/frame/benchmarking-cli/src/writer.rs
Comment (drahnr) [inline utils/frame/benchmarking-cli/src/writer.rs:476]: Is there a way to tune the insertion of `\n` between `(` and `&[`. A few pages down the reverse change is introduced.

File: utils/frame/remote-externalities/src/lib.rs
Comment (drahnr) [inline utils/frame/remote-externalities/src/lib.rs:273]: A few pages up the inverted change is applied.

File: utils/frame/rpc/system/src/lib.rs
Comment (drahnr) [inline utils/frame/rpc/system/src/lib.rs:87]: I find this rather unfortunate formatting and harder to read than the original.

**Review by drahnr — CHANGES_REQUESTED** [review]
As much as I would love this to land rather sooner than later, imho the fmt _introduction_ changeset should be as small as possible and a re-ordering imports could be defered. There are also a one or two patterns introduced i.e. `pattern => { .. },` which - if enforced by CI - could cause some transition friction.

File: utils/frame/benchmarking-cli/src/writer.rs
Comment (expenses) [reply utils/frame/benchmarking-cli/src/writer.rs:476]: This might be due to inconsistent formatting in the original code, but I'm not sure.

File: utils/build-script-utils/src/git.rs
Comment (coriolinus) [reply utils/build-script-utils/src/git.rs:37]: They're added because of the `match_block_trailing_comma = true` directive. I'd personally prefer to not set that directive, but I don't feel very strongly either way.

File: utils/frame/benchmarking-cli/src/writer.rs
Comment (coriolinus) [reply utils/frame/benchmarking-cli/src/writer.rs:476]: I don't think that the formatting of the original code matters much to rustfmt. I suspect this is just an artifact of its line-packing heuristics acting up. 

```toml
overflow_delimited_expr = true
```

might improve the situation though.

**Review by coriolinus — APPROVED** [review]
I strongly support adding a standardized rustfmt configuration. Enabling auto-formatting should improve productivity because nobody has to _think_ about formatting anymore. I also like the idea of eventually enforcing formatting with CI, though that's probably better as a follow-up.

No matter what, enabling rustfmt is going to result in a Really Big Diff. My take is that it's better to bite the bullet and incorporate all the changes at once, instead of dribbling in features like `reorder_imports` slowly over multiple PRs.

We may want to consider adding

```toml
license_template_path = "HEADER-APACHE2"
```

which would cause rustfmt to complain if someone forgets to add an appropriate license header. That's not essential though. 

Overall, I'm less concerned with _which_ particular style we adopt, and more interested in getting _some_ machine-enforced style rules into use.

File: rustfmt.toml
Comment (expenses) [reply rustfmt.toml:1]: We need a `rustfmt.toml` formatter 😅

File: rustfmt.toml
Comment (expenses) [reply rustfmt.toml:3]: I think it's better to be explicit here.

File: rustfmt.toml
Comment (expenses) [reply rustfmt.toml:4]: I've removed this. wrapping comments messes up some things like diagrams in comments.

File: utils/frame/benchmarking-cli/src/writer.rs
Comment (expenses) [reply utils/frame/benchmarking-cli/src/writer.rs:478]: I didn't see an improvement when I tried this.

File: utils/build-script-utils/src/git.rs
Comment (expenses) [reply utils/build-script-utils/src/git.rs:37]: Removed this.

**Comment by expenses** [comment]
> No matter what, enabling rustfmt is going to result in a Really Big Diff. My take is that it's better to bite the bullet and incorporate all the changes at once, instead of dribbling in features like `reorder_imports` slowly over multiple PRs.

My thinking here is that by doing seperate PRs for client/primitives/etc, I'll be able to get more reviewers and get people to add `#[rustfmt::skip]` to things that really need to be formatted a certain way.

**Comment by andresilva** [comment]
This needs approval from @gavofyork.

File: rustfmt.toml
Comment (gilescope) [reply rustfmt.toml:6]: The reordering of imports does not cross empty lines so you can still group std and others for example.

File: utils/browser/src/lib.rs
Comment (gilescope) [reply utils/browser/src/lib.rs:176]: If the one liner is too long we can reduce max line length?

File: rustfmt.toml
Comment (gilescope) [reply rustfmt.toml:1]: Formatting toml files is this issue: https://github.com/rust-lang/rustfmt/issues/4091

File: utils/build-script-utils/src/lib.rs
Comment (gilescope) [reply utils/build-script-utils/src/lib.rs:21]: better to have one hit of pain/disruption than several waves of disruption. As someone else pointed out the diff will be XXL either way.

File: utils/frame/rpc/system/src/lib.rs
Comment (gilescope) [reply utils/frame/rpc/system/src/lib.rs:87]: Could try struct_lit_single_line=false ? ( https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#struct_lit_single_line )

File: utils/build-script-utils/src/lib.rs
Comment (expenses) [reply utils/build-script-utils/src/lib.rs:21]: > better to have one hit of pain/disruption than several waves of disruption. As someone else pointed out the diff will be XXL either way.


Yeah fair.

File: utils/frame/rpc/system/src/lib.rs
Comment (expenses) [reply utils/frame/rpc/system/src/lib.rs:87]: This happened when I tried that:

Pretty unfortunate 😬 
```diff
diff --git a/utils/fork-tree/src/lib.rs b/utils/fork-tree/src/lib.rs
index 3d5cc85722..0a140326fe 100644
--- a/utils/fork-tree/src/lib.rs
+++ b/utils/fork-tree/src/lib.rs
@@ -1195,10 +1219,42 @@ mod test {
 				}
 			};
 
-			tree.import("A0", 1, Change { effective: 5 }, &is_descendent_of).unwrap();
-			tree.import("A1", 1, Change { effective: 5 }, &is_descendent_of).unwrap();
-			tree.import("D", 10, Change { effective: 10 }, &is_descendent_of).unwrap();
-			tree.import("E", 15, Change { effective: 50 }, &is_descendent_of).unwrap();
+			tree.import(
+				"A0",
+				1,
+				Change {
+					effective: 5,
+				},
+				&is_descendent_of,
+			)
+			.unwrap();
+			tree.import(
+				"A1",
+				1,
+				Change {
+					effective: 5,
+				},
+				&is_descendent_of,
+			)
+			.unwrap();
+			tree.import(
+				"D",
+				10,
+				Change {
+					effective: 10,
+				},
+				&is_descendent_of,
+			)
+			.unwrap();
+			tree.import(
+				"E",
+				15,
+				Change {
+					effective: 50,
+				},
+				&is_descendent_of,
+			)
+			.unwrap();
 
 			(tree, is_descendent_of)
 		};
```

**Comment by expenses** [comment]
Okay, I'd like to have a decision made on this. Either we accept rustfmt despite the fact that there are some style nitpicks here and there or we reject it, either temporarily (until we can get some changes into rustfmt) or forever.

**Review by coriolinus — APPROVED** [review]
I still think this is a really good idea. I also agree that we need approval from @gavofyork before proceeding.

**Comment by gilescope** [comment]
I think we need to have a second PR trying to apply this rustfmt to the most evil where clauses we can find. I'd suggest that client/conensus/* has some quite expansive where clauses... (unless someone can suggest something even trickier for rustfmt?)

**Comment by gilescope** [comment]
I've rustfmt consensus/* using the latest rules here: https://github.com/paritytech/substrate/pull/9350
- pretty impressed with the results.

**Comment by gilescope** [comment]
spaces_around_ranges = true - do we really want this?
```
for idx in 0 .. self.database_size.keys() {
    actual_warmup_key[16 ..].copy_from_slice(&kv.0[16 ..]);
}			
```
versus the default:
```
for idx in 0..self.database_size.keys() {
    actual_warmup_key[16..].copy_from_slice(&kv.0[16 ..]);
}			
```
(please thums up / thumbs down emoji vote below)

**Comment by thiolliere** [comment]
I'm in favor of this, I find the formatting good, it will also prevent people from having unrelated format change in their PR

**Comment by gilescope** [comment]
Not sure about this one:
`group_imports = "StdExternalCrate"`

If it's selected then a new line doesn't stop the grouping so the below would get merged into one:

```rust
use sp_inherents::{InherentData, CheckInherentsResult};
use static_assertions::const_assert;
use pallet_contracts::weights::WeightInfo;
use pallet_election_provider_multi_phase::FallbackStrategy;

#[cfg(any(feature = "std", test))]
pub use sp_runtime::BuildStorage;
#[cfg(any(feature = "std", test))]
pub use pallet_balances::Call as BalancesCall;
```

I think we're better just having the default which will group/sort but only up to a newline.

**Comment by gilescope** [comment]
`reorder_impl_items` will alphabetically order the impl items. I don't think we should mandate that - there's pleanty of times we might be grouping things together that are not lexically ordered. I think it would get in the way and be surprising to people more than it would help.

**Comment by thiolliere** [comment]
at the same time the prefix of crates are usually quite well organized: frame_*, pallet_*, sp_* ...
I'm a bit surprise `pub use` and `use` are grouped together but I don't mind.

**Comment by seunlanlege** [comment]
+1 for this change, code formatting should be added to ci

**Comment by gilescope** [comment]
Well with one new line they can be separated - we keep some control (best of both worlds?).

It's not a deal breaker but `match_arm_blocks=false` tightens things up a little (though it does look a little odd if the line above is indented - I've raised this with them).

File: utils/browser/src/lib.rs
Comment (gilescope) [reply utils/browser/src/lib.rs:157]: overflow_delimited_expr can help sometimes in these situations but on balence I think the code works better left to the default value.

File: utils/frame/rpc/system/src/lib.rs
Comment (gilescope) [reply utils/frame/rpc/system/src/lib.rs:87]: Idk, having it all on one line seems fine as long as it's not overly long. When it does get too long it breaks it down to one arg per line. There's a break even point where one is preferable to the other - I think we just have to agree on max line length.

**Comment by gilescope** [comment]
On the subject of trailing commas it looks a bit cleaner without them. (With trailing commas I do see occasionally trailing commas on arguments which looks very odd).

**Comment by gilescope** [comment]
Ok we have a _lot_of code. I've eyballed 1/4 of substrate now and am extreamly impressed with rustfmt. It's handled the where clauses well and in many cases formats them better than we did. In fact there are many many places where rustfmt is making improvements in readability and consistency. I was going to say that for the 1% of cases where it's not able to do a good job we can use rustfmt::skip but frankly rustfmt seems to do a reasonable job 100% of the time. In the 1/4 of the codebase I've looked at there's not been a case where I've felt I needed to jump in because rustfmt buchered it. Considering the codebase size it's testiment to the sweat and tears of the rustfmt team that they've managed to make rustfmt so good and able to handle all the edge cases.

The only thing I can really fault them for is line breaks before accessing a tuple member (I've raised that, but I don't think we should wait for that given how insignificant that is in our codebase as we (rightly) prefer named structs).

rustfmt has come of age.

**Comment by gilescope** [comment]
(btw the standard we should be holding rustfmt to is not "would I format my code it like that?" but rather "would I approve a PR formatted like that" as we only ourselves code a subset of the whole codebase)

File: rustfmt.toml
Comment (gilescope) [reply rustfmt.toml]: I think expenses has integrated all these suggestions.

**Comment by tomusdrw** [comment]
+1 for any semi-automated help with formatting. Currently doing everything manually is tedious. I don't have any strong preferences about any of the rules set (like honestly I don't care), but I just like the idea that the editor can take some mental overhead of formatting the code off me.
Sure, it's inevitable that some aesthetic trade-off will need to be accepted, but IMHO overall it's much more beneficial for both internal and external contributions to focus on the code correctness and quality instead of adding commas and new lines here and there.

File: rustfmt.toml
Comment (gilescope) [inline rustfmt.toml:17]: I think this set of changes gets us to compliance with the stype guide ( https://github.com/paritytech/substrate/blob/master/docs/STYLE_GUIDE.md ) with the exception that the 'where' block is not indented by one level. I think this is something we can get used to and we might appreciate the additional width given the size of some of our where clauses.

```suggestion
imports_granularity = "Crate"
reorder_imports = true
# Consistency
newline_style = "Unix"
normalize_comments = true
normalize_doc_attributes = true
# Misc
chain_width = 80
spaces_around_ranges = false
binop_separator = "Back"
reorder_impl_items = false
match_arm_leading_pipes = "Preserve"
match_arm_blocks = false
match_block_trailing_comma = true
trailing_comma = "Vertical"
trailing_semicolon = false
```

**Comment by andresilva** [comment]
This gets a 👍 from me. To be honest I don't really care about the specific formatting style we end up with, and therefore will just let people with stronger opinions lead this, although I think that making it as close as possible to the existing style (which is not very well defined) is probably the way to go. For me the main advantage of using rustfmt is consistency and not having to care about formatting anymore, any second that someone spends thinking about code style when writing code or reviewing is a wasted second in my opinion.

File: utils/frame/rpc/system/src/lib.rs
Comment (chevdor) [reply utils/frame/rpc/system/src/lib.rs:87]: In general, I am totally in favor of having (any) kind of standard that can be applied automatically. This is a time & brain saver.

The example above is probably very marginal but indeed annyoing. There are times where you have data (even long lines) and it is way more readable to keep it on single lines.

The issue above seem addressable using `struct_lit_single_line = true`.

Alternatively, I think this can also be shutdown for those exceptions with:
```
#[rustfmt::skip]
{
  tree.import("A0", 1, Change { effective: 5 }, &is_descendent_of).unwrap();
  tree.import("A1", 1, Change { effective: 5 }, &is_descendent_of).unwrap();
  tree.import("D", 10, Change { effective: 10 }, &is_descendent_of).unwrap();
  tree.import("E", 15, Change { effective: 50 }, &is_descendent_of).unwrap();
}
```

It works however it changes the meaning of the code.

**Comment by gilescope** [comment]
On the CI we should also set `license_template_path` so that we automatically check the license put at the top of all our .rs files rather than manually commenting in a PR that it's missing.

File: utils/browser/src/lib.rs
Comment (chevdor) [reply utils/browser/src/lib.rs:158]: I think the original is indeed better readable here.
Ideally would be a rule to:
- keep `(..).map(drop),` as such
- while letting `(..).map(drop).foo(bar)` gets split eventually, whether always or dependending on the line width (I would be fine with both options eitherway)

**Review by gavofyork — APPROVED** [review]

**Comment by kianenigma** [comment]
I highly recommend 

```
comment_width = 100
wrap_comments = true
format_strings = true
```

as well to clean up all of our docstring (which is totally arbtirary length atm).

**Comment by expenses** [comment]
> I highly recommend
> 
> ```
> comment_width = 100
> wrap_comments = true
> format_strings = true
> ```
> 
> as well to clean up all of our docstring (which is totally arbitrary-length atm).

I'd like to have these as well, but they trip up occasionally:

```diff
diff --git a/utils/wasm-builder/src/builder.rs b/utils/wasm-builder/src/builder.rs
index 20f33583b8..1d6f936fe3 100644
--- a/utils/wasm-builder/src/builder.rs
+++ b/utils/wasm-builder/src/builder.rs
@@ -195,8 +195,8 @@ fn provide_dummy_wasm_binary_if_not_exist(file_path: &Path) {
 	if !file_path.exists() {
 		crate::write_file_if_changed(
 			file_path,
-			"pub const WASM_BINARY: Option<&[u8]> = None;\
-			 pub const WASM_BINARY_BLOATY: Option<&[u8]> = None;",
+			"pub const WASM_BINARY: Option<&[u8]> = None;pub const WASM_BINARY_BLOATY: \
+			 Option<&[u8]> = None;",
 		);
 	}
 }
```

Would it be acceptable to save these changes for a future PR?

**Comment by bkchr** [comment]
> Would it be acceptable to save these changes for a future PR?

Why? 

The changes above look weird. Not sure we really need `format_strings`.

**Comment by expenses** [comment]
> > Would it be acceptable to save these changes for a future PR?
> 
> Why?
> 
> The changes above look weird. Not sure we really need `format_strings`.

We might find a way for them to not be weird. But it doesn't matter, we won't include them for now.

**Review by drahnr — APPROVED** [review]

**Comment by chevdor** [comment]
@expenses I think we can find small details to fix for such a PR pretty much forever :) I also added a comment but overall, I find the result very positive.

Of course, when it comes to formattings, the merging of this PR will generate extra work (potentially) subsequent changes to the standard are a bit of a pain, especially for open PRs that will need to rebase and fix a bunch of conflicts. 

In short, if we can be 90% happy (and that's the case AFAIC) with the current PR, I think we should freeze its state and share the agreed config for open PRs to *get ready*.

I would recommend to:
- freeze the state of the PR
- announce the upcoming change and let PR authors merge or prepare big PRs before we merge this one

**Review by chevdor — APPROVED** [review]

**Comment by TriplEight** [comment]
Do you guys want it as a blocking check in CI or is it good enough, for now, to have it launched locally?
Normally it's both.

**Comment by expenses** [comment]
Okay great. Yeah @chevdor that's a good idea. I think that with the number of approvals we've got so far, we can go ahead and at least commit the `rustfmt.toml`. We can actually format the code in other PRs, as it only requires a single `cargo +nightly fmt`.

**Comment by kianenigma** [comment]
> ```
> format_strings = true
> ```

Then let's get rid of `format_strings = true`, the other two pretty good IMO.

**Comment by expenses** [comment]
> Do you guys want it as a blocking check in CI or is it good enough, for now, to have it launched locally?
> Normally it's both.

We only want this in CI once we've done the initial formatting of the code.

**Comment by expenses** [comment]
bot merge

**Comment by drahnr** [comment]
Quickly attempt to use the used toml in substrate, filed this upstream issue for `normaliz_comments` https://github.com/rust-lang/rustfmt/issues/4909

**Pull request merged**
